# Précisions
N'oubliez pas d'aller consulter les ressources sur les fonctions existentes de pandas, numpy et seaborn qui vous seront utiles lorsque vous réponderez aux questions qui suivent. Voici les sites web concernés à consulter pour les plus récentes versions de chacunes des librairies scientifiques:
* Pandas: [documentation](https://pandas.pydata.org/docs/)
* Numpy: [documentation](https://numpy.org/doc/1.21/)
* Matplotlib: [quelques fonctions plot() pertinentes](https://matplotlib.org/3.5.0/plot_types/index.html)
* Seaborn: [introduction à seaborn](https://seaborn.pydata.org/introduction.html)

Tout au long du TP, vous pouvez cliquer sur les liens bleus qui vous mèneront vers des ressources encore plus spécifiques pour réaliser chacune des questions.

# Partie 1: Exploration de la base de données

In [3]:
#Différents 'imports' nécessaires pour ce TP
import os
import numpy as np
import pandas as pd 
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib
matplotlib.rcParams['figure.figsize'] = [12,6] 
plt.style.use('fivethirtyeight')

## 1.1 Visualisation des données

Nous allons travailler avec un fichier .csv (comma separated values), qui contient les 'Happiness score', soit les indicateurs de bonheur de l'année 2020 pour 153 pays du monde. Ces pays sont organisés sur les rangées du Dataframe 'df', une structure de données tabulaire de la librairie pandas. Des paramètres additionnels sont disponibles sur chaque colonnes, qui incluent notamment l'indicateur régional ('Region'), le support social ('Social Support'), le logarithme du produit intérieur brut par capita ('Logged GDP per capita') et plusieurs autres. Il faut en premier visualiser les données en notre possession. Pour ce faire, rouler le code qui suit. 

In [4]:
#Trouve le chemin relatif pour accéder aux données par la suite
path = os.getcwd()

df = pd.read_csv(os.path.join(path, '2020.csv'))

#Précision mise à 2 chiffres après la virgule
pd.set_option("display.precision", 2)

#df.head(x) permet d'afficher les x premières rangées de données, cette fonction vous sera utile pour la suite des choses
#lors de la visualisation de la banque de données modifiée, utilisez la fonction display() au lieu de print() dans le 
#jupyter notebook pour un meilleur affichage
display(df.head(15))
# df.describe()

#La commande ci-dessous vous donne de l'information sur la base de données (dataset), la quantité de valeurs et le type
#de variable utilisée (dtypes, à droite complètement ex:float64 et object est un string)
df.info()
print('\n')
#Pour retrouver un élément de la case 'Country name'
print(f"Le premier élément de la colonne Country name est : {df['Country name'][0]}")

#Nous pouvons aussi retrouver le nombre de régions en utilisant la ligne de code suivante:
df['Region'].value_counts()

Country name                        Region  Happiness score  \
0          Finland                Western Europe             7.81   
1          Denmark                Western Europe             7.65   
2      Switzerland                Western Europe             7.56   
3          Iceland                Western Europe             7.50   
4           Norway                Western Europe             7.49   
5      Netherlands                Western Europe             7.45   
6           Sweden                Western Europe             7.35   
7      New Zealand         North America and ANZ             7.30   
8          Austria                Western Europe             7.29   
9       Luxembourg                Western Europe             7.24   
10          Canada         North America and ANZ             7.23   
11       Australia         North America and ANZ             7.22   
12  United Kingdom                Western Europe             7.16   
13          Israel  Middle East and North Africa             7.13   
14      Costa Rica   Latin America and Caribbean             7.12   

    Standard error of happiness score  Upper Confidence Interval  \
0                                0.03                       7.87   
1                                0.03                       7.71   
2                                0.04                       7.63   
3                                0.06                       7.62   
4                                0.03                       7.56   
5                                0.03                       7.50   
6                                0.04                       7.42   
7                                0.04                       7.38   
8                                0.03                       7.36   
9                                0.03                       7.30   
10                               0.04                       7.31   
11                               0.04                       7.30   
12                               0.04                       7.24   
13                               0.04                       7.21   
14                               0.04                       7.21   

    Lower Confidence Interval  Logged GDP per capita  Social support  \
0                        7.75                  10.64            0.95   
1                        7.58                  10.77            0.96   
2                        7.49                  10.98            0.94   
3                        7.39                  10.77            0.97   
4                        7.42                  11.09            0.95   
5                        7.39                  10.81            0.94   
6                        7.28                  10.76            0.93   
7                        7.22                  10.50            0.95   
8                        7.23                  10.74            0.93   
9                        7.18                  11.45            0.91   
10                       7.15                  10.69            0.93   
11                       7.14                  10.72            0.94   
12                       7.09                  10.60            0.94   
13                       7.05                  10.42            0.91   
14                       7.03                   9.66            0.90   

    Healthy life expectancy  Freedom to make life choices  Generosity  \
0                      71.9                          0.95   -5.95e-02   
1                      72.4                          0.95    6.62e-02   
2                      74.1                          0.92    1.06e-01   
3                      73.0                          0.95    2.47e-01   
4                      73.2                          0.96    1.35e-01   
5                      72.3                          0.91    2.08e-01   
6                      72.6                          0.94    1.12e-01   
7                      73.2                          0.94    1.92e-01   
8 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 12 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Country name                       153 non-null    object 
 1   Region                             153 non-null    object 
 2   Happiness score                    153 non-null    float64
 3   Standard error of happiness score  153 non-null    float64
 4   Upper Confidence Interval          153 non-null    float64
 5   Lower Confidence Interval          153 non-null    float64
 6   Logged GDP per capita              153 non-null    float64
 7   Social support                     153 non-null    float64
 8   Healthy life expectancy            153 non-null    float64
 9   Freedom to make life choices       153 non-null    float64
 10  Generosity                         153 non-null    float64
 11  Perceptions of corruption          153 non-null    float64

Sub-Saharan Africa                    39
Western Europe                        21
Latin America and Caribbean           21
Middle East and North Africa          17
Central and Eastern Europe            17
Commonwealth of Independent States    12
Southeast Asia                         9
South Asia                             7
East Asia                              6
North America and ANZ                  4
Name: Region, dtype: int64

## Question 1.
Pour cette question, vous devez retourner les valeurs maximales de 'Happiness score' pour chacune des 10 régions ('Region'). Ainsi la fonction ci-dessous doit retourner un dictionnaire avec ces scores maximaux en ordre décroissant et leurs pays associés en tant que clé. Par exemple: Australia:7.3, Israel:7.21, etc. N'oubliez pas que vous avez accès aux librairies de pandas et numpy qui possèdent plein d'outils pour automatiser un processus de ce genre!

### 1.1 
Vous devez en premier lieu créer une fonction simple qui retourne une base de données de la région voulue, soit celle donnée en paramètre à la fonction CreateSubsetPerRegion(). Un exemple de ce qui doit être retourné est présenté ci-dessous:
![](Capture1.PNG)

In [5]:
def CreateSubsetPerRegion(df, region):
    #TODO Extraire les sous-données par région à l'aide du dataframe pandas ('subset' de données)
    subset = df[df["Region"] == region]
    return subset

display(CreateSubsetPerRegion(df,'South Asia'))

Country name      Region  Happiness score  \
65      Pakistan  South Asia             5.69   
86      Maldives  South Asia             5.20   
91         Nepal  South Asia             5.14   
106   Bangladesh  South Asia             4.83   
129    Sri Lanka  South Asia             4.33   
143        India  South Asia             3.57   
152  Afghanistan  South Asia             2.57   

     Standard error of happiness score  Upper Confidence Interval  \
65                                0.05                       5.79   
86                                0.07                       5.34   
91                                0.06                       5.25   
106                               0.04                       4.91   
129                               0.05                       4.43   
143                               0.03                       3.63   
152                               0.03                       2.63   

     Lower Confidence Interval  Logged GDP per capita  Social support  \
65                        5.60                   8.48            0.69   
86                        5.06                   9.52            0.91   
91                        5.02                   7.93            0.79   
106                       4.75                   8.29            0.69   
129                       4.22                   9.39            0.83   
143                       3.52                   8.85            0.59   
152                       2.51                   7.46            0.47   

     Healthy life expectancy  Freedom to make life choices  Generosity  \
65                     58.25                          0.73        0.04   
86                     70.60                          0.85        0.04   
91                     63.78                          0.80        0.15   
106                    64.50                          0.90       -0.03   
129                    67.20                          0.84        0.08   
143                    60.22                          0.88        0.06   
152                    52.59                          0.40       -0.10   

     Perceptions of corruption  
65                        0.75  
86                        0.82  
91                        0.74  
106                       0.66  
129                       0.86  
143                       0.77  
152                       0.93

### 1.2
Il faut maintenant retourner le dictionnaire ordonné en ordre décroissant du pays associé à la valeur maximale par région tel qu'indiqué dans la question 1. À nouveau, n'oubliez pas que vous avez accès à des librairies qui automatisent ces processus plus facilement. 

In [24]:
def ClassMaxScorePerRegion(df):
    #TODO Trouver les régions pour en faire une liste (https://numpy.org/doc/stable/reference/generated/numpy.unique.html)
    regions = np.unique(df["Region"])
    #TODO Pour chacune des régions, trouves les valeurs maximales et le pays associé
    dictionnaire = {}
    for region in regions:
        maximum = pd.DataFrame.max(df[df["Region"] == region]["Happiness score"])
        ligne = df[df["Happiness score"] == maximum]
        pays = ligne.iloc[0,0]  
        dictionnaire[pays] = maximum
    #TODO Création d'un dictionnaire en ordre décroissant des valeurs de 'Happiness score' 
    sorted_dict = dict(sorted(dictionnaire.items(),reverse = True,key= lambda x:x[1]))  
    return sorted_dict

print(ClassMaxScorePerRegion(df))


{'Finland': 7.808700085, 'New Zealand': 7.299600124, 'Israel': 7.128600121, 'Costa Rica': 7.121399879, 'Czech Republic': 6.910900116, 'Taiwan': 6.45539999, 'Singapore': 6.377099991, 'Uzbekistan': 6.257599831, 'Mauritius': 6.101299763, 'Pakistan': 5.69329977}


## Question 2
Nous allons maintenant rajouter les données de 2016 afin de les comparer à celles de 2020. Tout d'abord, utilisez ImportNewData() pour importer cette nouvelle base de données nommée 2016.csv. Avec les pays ayant obtenu les scores maximaux de 'Happiness score' par région, le but est de tracer leur évolution entre 2016 à 2020. L'évolution doit être représentée de la plus petite différence à la plus grande. Pour ce faire, vous allez en premier lieu trouver les différences avec la fonction FindDifferences() qui retourne un dictionnaire en ordre croissant de ces différences. Par la suite, vous tracerez ces différences par pays. Voici un exemple de graphique à barre voulu à réaliser avec la fonction PlotComparison(), ce ne sont pas nécessairement les pays voulus. Vous pouvez vous renseigner sur la fonction [plt.bar](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.bar.html) de matplotlib.
![](Capture2.PNG)

In [22]:
def ImportNewData():
    #TODO Import de la base de données pour 2016
    #TODO Afficher les 10 premières rangées de la base de données de 2016
    path = os.getcwd()

    df_2016 = pd.read_csv(os.path.join(path, '2016.csv'))

    
    return df_2016
    
df_2016 = ImportNewData()

In [36]:
def FindDifferences(df_1,df_2):
    #TODO Reprendre les valeurs obtenues précédemment du pays avec le score maximal par région
    #TODO Retourner un dictionnaire des différences en ordre croissant
    valeurs = ClassMaxScorePerRegion(df)
    diff = {}
    for key in valeurs:
        ancienne = np.ndarray.tolist(df_2[df_2["Country name"] == key]["Happiness score"])[0]
        diff[key] = valeurs[key] - ancienne
    sorted_dict = dict(sorted(diff.items(),key= lambda x:x[1])) 
    return sorted_dict

def PlotComparison(sorted_dict):
    #TODO Faire un bar plot de ces différences
    pays = list(sorted_dict.keys())
    diff = list(sorted_dict.values())

    plt.bar(pays,diff)
    plt.xlabel("Pays")
    plt.ylabel("Différence de score de bonheur")
    plt.title("Difference entre 2016 et 2020 des scores de bonheur en fonction du pays")
    plt.show()
    
sorted_dict = FindDifferences(df, df_2016)
PlotComparison(sorted_dict)


TypeError: descriptor 'tolist' for 'numpy.ndarray' objects doesn't apply to a 'Series' object

## Question 3
Seaborn est une librairie qui permet de plus facilement tracer des graphiques complexes. Souvent, des [subplot](https://matplotlib.org/stable/gallery/subplots_axes_and_figures/subplots_demo.html) via matplotlib sont utilisés afin d'afficher plusieurs graphiques dans la même figure à des fins de comparaisons. Nous nous intéresserons maintenant à deux paramètres particuliers:
1. Healthy life expectancy
2. Happiness score

Vous devez comparer pour les deux années 2016 et 2020 ces deux paramètres pour un pays et une région donnés en paramètre de la fonction PlotDistributions() ci-dessous. Pour le premier paramètre, vous utiliserez [sns.displot](https://seaborn.pydata.org/generated/seaborn.distplot.html) et pour le second paramètre, vous ferez usage des fonctionnalités de pandas pour le traçage avec [boxplot](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.boxplot.html). Assurez vous également que le pays donné en paramètre de la fonction à écrire correspond à la région associée avant de tracer les graphiques voulus. Voici un exemple d'affichage:
![](Capture3.PNG)


En cas de région et de pays qui ne correspondent pas, vous devez afficher:
![](Capture4.PNG)

In [ ]:
def PlotDistributions(df_1, df_2, region, country):
    #TODO  Vérifier si le pays fait partie de la région donnée sinon afficher le message indiqué plus haut
        
    #TODO Utiliser sns.displot et df.boxplot pour visualiser les deux paramètres demandés
        

PlotDistributions(df_2016, df, 'North America and ANZ', 'Canada')
PlotDistributions(df_2016, df, 'Western Europe', 'Taiwan')

## Question 4
En se concentrant maintenant à nouveau sur les données de **2020**, nous allons essayer de trouver le paramètre le plus corrélé avec le score de bonheur 'Happiness score'. Nous allons prendre des paramètres spécifiques à analyser:
1. Happiness score
2. Logged GDP per capita
3. Social support
4. Healthy life expectancy
5. Freedom to make life choices
6. Generosity
7. Perceptions of corruption

Utilisez un [sns.heatmap](https://seaborn.pydata.org/generated/seaborn.heatmap.html) afin de visualiser ces [corrélations](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.corr.html). Votre heatmap de corrélation devrait ressembler à ce qui est présenté ci-dessous. Attention, ce qui est affiché n'est pas exactement ce que vous allez obtenir puisque ce ne sont pas les même données utilisées. L'exemple n'est qu'indicatif pour la ressemblance visuelle à obtenir.
![](Capture5.PNG)

Pour ce faire, vous allez écrire la fonction FindCorrelations(). En plus d'afficher le heatmap, cette fonction retourne la valeur maximale de corrélation avec le score de bonheur en considérant un paramètre différent de lui-même. En effet, comme vous pouvez le voir à la figure ci-haut le paramètre 'Happiness score' est toujours corrélé à 100% avec lui-même. Nous nous intéresserons donc plutôt à la deuxième valeur de corrélation la plus élevée et le paramètre qui lui est associé. L'affichage final ressemble à ceci et le coefficient de corrélation doit être arrondi au centième près:
![](Capture7.PNG)

In [ ]:
# Réajuste le fond d'affichage à blanc
matplotlib.rcParams['figure.facecolor'] = 'w'
def FindCorrelations(df):
    #TODO Sélectionner les 7 paramètres à étudier
    #TODO Faire et afficher un heatmap des facteurs de corrélation
    
    plt.show()
    return chosen_max_cor, param
    
chosen_max_cor, param = FindCorrelations(df)
#TODO Afficher la phrase indiquant le coefficient de corrélation le plus élevé avec le paramètre 'Happiness score'
print(...)

## Suite...
Rendez-vous à la partie 3 que vous pouvez retrouver dans le fichier tests.py du github et dont les instructions sont dans le readme.